### PR TITLE
DAOS-623 cq: simplify githook output

### PR DESCRIPTION
**Pull request by daltonbohning**
Skip-test: true
Skip-tests: true
Skip-unit-tests: true
Skip-fault-injection-test: true

Add _print_githook_header to align githook header output. Add _git_diff_cached_files so hooks get a diff the same way. Remove most verbose output from githooks.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by daltonbohning** [comment]
Example output when using `git commit`. More improvements could be made
```
Checking for target branch
  Using origin as remote repo.  If this is incorrect, set DAOS_ORIGIN in environment
  Install gh command to auto-detect target branch, assuming origin/master.
Copyright:        Updating copyright headers
CodeSpell:        Checking for spelling mistakes
Jenkinsfile:      No Jenkinsfile changes. Skipping
Yaml Lint:        Checking yaml formatting
Clang-format:     git-clang-format not installed. Skipping
Gofmt:            No GO changes. Skipping
isort:            Checking if python imports are sorted
Skipped 6 files
Flake8:           Linting python
Pylint:           Linting python
FileTypeList

Checking for target branch
  Using origin as remote repo.  If this is incorrect, set DAOS_ORIGIN in environment
  Install gh command to auto-detect target branch, assuming origin/master.
Checking for target branch
  Using origin as remote repo.  If this is incorrect, set DAOS_ORIGIN in environment
  Install gh command to auto-detect target branch, assuming origin/master.
Checking that required hooks ran
```

File: utils/githooks/hook_base.sh [hunk 1]
```diff
@@ -1,9 +1,33 @@
...
 set -eu
 
 . utils/githooks/find_base.sh
 export TARGET
 
+# Common function to keep headers aligned
+function _print_githook_header() {
+    printf "%-17s " "${1}:"
+}
+export -f _print_githook_header
```
[the remark this answers is not in the document]
Comment (daltonbohning) [reply utils/githooks/hook_base.sh:16]: Also, should these be "private" / start with `_`?

File: utils/githooks/pre-commit.d/10-update-copyright [hunk 2]
```diff
@@ -39,20 +45,16 @@ targets=(
     '.env'
 )
 
+if [ -z "$files" ]; then
+    files=$(git diff "$TARGET" --cached --diff-filter=AM --name-only -- "${targets[@]}")
+else
+    echo "  Checking against custom files"
+fi
```
Comment (daltonbohning) [inline utils/githooks/pre-commit.d/10-update-copyright:52]: I don't see where `files` would be set. Presumable in someone's ENV? Maybe we could just remove this handling now

File: utils/githooks/pre-commit.d/10-update-copyright [hunk 3]
```diff
@@ -39,20 +45,16 @@ targets=(
...
-if [ -z "$files" ]; then
-    if [ "$TARGET" = "HEAD" ]; then
-        echo "  Checking against HEAD"
-        files=$(git diff HEAD  --diff-filter=AM --name-only -- "${targets[@]}")
-    else
-        echo "  Checking against branch ${TARGET}"
-        files=$(git diff "$TARGET"... --diff-filter=AM --name-only -- "${targets[@]}")
-    fi
-fi
-
```
Comment (daltonbohning) [inline utils/githooks/pre-commit.d/10-update-copyright:55]: Since `TARGET` is always ran through `git merge-base` we always get a SHA here.

File: utils/githooks/pre-commit.d/20-codespell.sh [hunk 4]
```diff
@@ -9,20 +9,15 @@
...
     exit 0
 fi
 
-if [ "$TARGET" = "HEAD" ]; then
-        echo "  Checking against HEAD"
-        git diff HEAD --name-only | xargs codespell
-else
-        echo "  Checking against branch ${TARGET}"
-        git diff "$TARGET"... --name-only | xargs codespell
-fi
```
Comment (daltonbohning) [inline utils/githooks/pre-commit.d/20-codespell.sh:28]: Since `TARGET` is always ran through `git merge-base` we always get a SHA here.

**Comment by daltonbohning** [comment]
See [this comment](https://github.com/daos-stack/daos/pull/14214#issuecomment-2072880503) for example output.

File: utils/githooks/hook_base.sh [hunk 5]
```diff
@@ -1,9 +1,33 @@
...
+# shellcheck disable=SC2086
+function _git_diff_cached_files() {
+    IFS=' ' read -r -a _filter <<< "${1:-}"
+    local args="${2:-}"
+    if [ ${#_filter[@]} -eq 0 ]; then
+        git diff "$TARGET" --cached --name-only --diff-filter=d $args
+    else
+        git diff "$TARGET" --cached --name-only --diff-filter=d $args -- "${_filter[@]}"
+    fi
+}
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply utils/githooks/hook_base.sh:28]: Good work.

File: utils/githooks/pre-commit.d/70-isort.sh [hunk 6]
```diff
@@ -1,25 +1,36 @@
...
 
-echo "isort:"
+echo "Checking if python imports are sorted"
 
-echo "  Running isort for python imports."
-isort_args=(--jobs 8 .)
-if ! isort --check-only "${isort_args[@]}"; then
-    echo "  isort check failed, run 'isort ${isort_args[*]}' to fix."
+if ! echo "$py_files" | xargs -r isort --check-only --jobs 8; then
+    echo "  isort check failed, run 'isort --jobs 8 .' to fix."
```
Comment (daltonbohning) [inline utils/githooks/pre-commit.d/70-isort.sh:34]: isort was previously running against the entire source tree even if no python changes were made. On my boro node, this took ~1.5s

File: utils/githooks/pre-commit.d/70-isort.sh [hunk 7]
```diff
@@ -10,6 +10,13 @@ set -ue
 
 _print_githook_header "isort"
 
+py_files=$(_git_diff_cached_files "*.py SConstruct *SConscript")
```
Comment (ashleypittman) [inline utils/githooks/pre-commit.d/70-isort.sh:13]: what does `*SConscript` mean here? Is that matching subdirs or filenames?  All files should be called `SConscript` that I'm aware of.

Comment (daltonbohning) [reply utils/githooks/pre-commit.d/70-isort.sh:13]: It's to match subdirs. Maybe I should do `*/SConscript`?

**Review by phender — APPROVED** [review]

File: utils/githooks/pre-commit.d/10-update-copyright [hunk 8]
```diff
@@ -39,20 +45,16 @@ targets=(
     '.env'
 )
 
+if [ -z "$files" ]; then
+    files=$(git diff "$TARGET" --cached --diff-filter=AM --name-only -- "${targets[@]}")
```
Comment (brianjmurrell) [inline utils/githooks/pre-commit.d/10-update-copyright:49]: What are the ramifications of removing the `...` from target being diffed?

Comment (daltonbohning) [reply utils/githooks/pre-commit.d/10-update-copyright:49]: I tried to explain here
https://github.com/daos-stack/daos/pull/14214#discussion_r1581329370

TL;DR the `...` seems to only include changes that are not staged, or that are already committed.

By example
When a single file is modified and added:
```
$ git add Jenkinsfile
$ git status
Changes to be committed:
        modified:   Jenkinsfile

# Wrong because it doesn't included the staged file
$ git diff origin/master... --name-only
# empty!

# Correct in this case
$ git diff origin/master --name-only
Jenkinsfile
```

When one file is modified and added, and another file is modified but not added
```
$ git status
Changes to be committed:
        modified:   Jenkinsfile

Changes not staged for commit:
        modified:   utils/rpms/daos.spec

# Wrong because it doesn't include either file
$ git diff origin/master... --name-only
# empty!

# Technically wrong because daos.spec is not staged for commit.
# I.e. don't run commit hooks against files that are modified but not going to be committed.
$ git diff origin/master --name-only
Jenkinsfile
utils/rpms/daos.spec

# --cached includes just the staged files.
$ git diff origin/master --name-only --cached
Jenkinsfile
```

**Review by brianjmurrell — APPROVED** [review]

**Comment by daltonbohning** [comment]
Skipped CI test stages since only githooks were modified
